lndclient/versioner_client.go

```diff
@@ -0,0 +1,179 @@
+	// is lower than v0.10.0-beta.
+	ErrVersionCheckNotImplemented = errors.New("version check not " +
+		"implemented")
+
+	// ErrVersionIncompatible is the error that is returned if the connected
+	// lnd instance is not supported.
+	ErrVersionIncompatible = errors.New("version incompatible")
+
+	// ErrBuildTagsMissing is the error that is returned if the
+	// connected lnd instance does not have all built tags activated that
+	// are required.
+	ErrBuildTagsMissing = errors.New("build tags missing")
+)
+
+// VersionerClient exposes the version of lnd.
+type VersionerClient interface {
+	// GetVersion returns the version and build information of the lnd
+	// daemon.
+	GetVersion(ctx context.Context) (*verrpc.Version, error)
+}
+
+type versionerClient struct {
+	client      verrpc.VersionerClient
+	readonlyMac serializedMacaroon
+}
+
+func newVersionerClient(conn *grpc.ClientConn,
+	readonlyMac serializedMacaroon) *versionerClient {
+
+	return &versionerClient{
+		client:      verrpc.NewVersionerClient(conn),
+		readonlyMac: readonlyMac,
+	}
+}
+
+// GetVersion returns the version and build information of the lnd
+// daemon.
+//
+// NOTE: This method is part of the VersionerClient interface.
+func (v *versionerClient) GetVersion(ctx context.Context) (*verrpc.Version,
+	error) {
+
+	rpcCtx, cancel := context.WithTimeout(
+		v.readonlyMac.WithMacaroonAuth(ctx), rpcTimeout,
+	)
+	defer cancel()
+	return v.client.GetVersion(rpcCtx, &verrpc.VersionRequest{})
+}
+
+// CheckVersionCompatibility makes sure the connected lnd node has the correct
+// version and required build tags enabled.
+//
+// NOTE: This check will **never** return a non-nil error for a version of
+// lnd < 0.10.0 because any version previous to 0.10.0 doesn't have the version
+// endpoint implemented!
+func CheckVersionCompatibility(client VersionerClient, major, minor,
+	patch uint32, tags []string) error {
+
+	// First, test that the version RPC is even implemented.
+	version, err := client.GetVersion(context.Background())
+	if err != nil {
+		// The version service has only been added in lnd v0.10.0. If
+		// we get an unimplemented error, it means the lnd version is
+		// definitely older than that.
+		s, ok := status.FromError(err)
+		if ok && s.Code() == codes.Unimplemented {
+			return ErrVersionCheckNotImplemented
+		}
+		return fmt.Errorf("GetVersion error: %v", err)
+	}
+
+	// Now check the version and make sure all required build tags are set.
+	err = assertVersionCompatible(version, major, minor, patch)
+	if err != nil {
+		return err
+	}
+	err = assertBuildTagsEnabled(version, tags)
+	if err != nil {
+		return err
+	}
+
+	// All check positive, version is fully compatible.
+	return nil
+}
+
+// VersionString returns a nice, human readable string of a version returned by
+// the VersionerClient, including all build tags.
+func VersionString(version *verrpc.Version) string {
+	if version == nil {
+		return "<nil>"
```

Comment:
Or is this pre-0.10?

Reply:
I added this only because a test didn't specify a version in the mock. I fixed the test and removed the extra check.